File: service/GoodNight.Service.Domain/Model/Read/Transfer/Action.cs
using System.Collections.Immutable;

namespace GoodNight.Service.Domain.Model.Read.Transfer
{
  // Requirements stem from an expression of type bool
  public record Requirement(
    string Display,
    string? Icon,
    bool? Passed);

  // Tests stem from requirements with expressions of type chance
  public record Test(
    string Display,
    string? Icon,
    int Chance);

  public record Option(
    string Urlname,
    string Text,
    bool IsAvailable,
    IImmutableList<Requirement> Requirements,
    IImmutableList<Test> Tests);

  public record Action(
    string Text,
    IImmutableList<Property> Effects,
    IImmutableList<Option> Options,
    bool HasReturn,
    bool HasContinue);
}

File: client/src/state/model/read/Action.ts
import type {Property} from "./Property";


export interface Requirement {
  display: string;
  icon?: string;
  passed: boolean;
}

export interface Test {
  display: string;
  icon?: string;
  chance: number;
}


export interface Option {
  urlname: string;
  text: string;
  icon: string | null;
  isAvailable: boolean;
  requirements: Array<Requirement>;
  tests: Array<Test>;
}


export interface Action {
  text: string;
  effects: Array<Property>;
  options: Array<Option>;
  hasReturn: boolean;
  hasContinue: boolean;
}
